src/compute/src/render/join/delta_join.rs

```diff
@@ -324,9 +421,15 @@ fn build_halfjoin<G, Tr, CF>(
 where
     G: Scope,
     G::Timestamp: crate::render::RenderTimestamp,
-    Tr: TraceReader<Time = G::Timestamp, Key = Row, Val = Row, R = Diff> + Clone + 'static,
+    Tr: TraceReader<Time = G::Timestamp, Key = K, Val = V, R = Diff> + Clone + 'static,
+    K: ExchangeData + FromRowByTypes + Hashable + for<'a> IntoRowByTypes<'a>,
+    V: ExchangeData + for<'a> IntoRowByTypes<'a>,
     CF: Fn(&G::Timestamp, &G::Timestamp) -> bool + 'static,
 {
+    let updates_key_types = Vec::from(trace_val_types);
+    let trace_key_types = Vec::from(trace_key_types);
+    let trace_val_types = Vec::from(trace_val_types);
```

Comment:
Since you need to clone these into vecs in all cases, I think it's better to take them as vecs and clone them at the callsite if necessary.

Reply:
We are getting slices as parameters to make the call sites more consistent as per https://github.com/MaterializeInc/materialize/pull/21968#discussion_r1342825136. That way, when we eventually get to writing macros to generate some of the dispatch functions, we can potentially have an easier time.

Reply:
The [Rust API guidelines](https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control) agree with Brennan here. I don't think we need to follow them religiously though, if they aversely affect readability.

> That way, when we eventually get to writing macros to generate some of the dispatch functions, we can potentially have an easier time.

Can you explain what you mean by that? I don't see how passing in `Vec`s makes it harder to write macros than passing slices. Wouldn't you simply exchange `vec![<types>]` for `&[<types>]`?

Reply:
I just changed this occurrence for now to follow the guidelines and get owned `Vec`s as parameters; we can deal with how the macros should look like when we get to writing them in the future.